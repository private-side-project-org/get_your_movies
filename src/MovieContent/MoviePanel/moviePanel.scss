@use "assets/variables" as *;

.moviePanel {
  &-container {
    overflow: scroll;
    display: flex;
    flex-direction: column;
    padding: 16px;
    position: relative;
    max-width: 710px;

    .moviePanel-image-container {
      animation-name: moveRight;
      animation-duration: 0.5s;
      animation-timing-function: ease-in-out;
      max-width: 250px;
      width: 100%;
      display: flex;
      flex-direction: column;
      margin-bottom: 24px;

      > img {
        border-radius: 6px;
      }

      @include mq(xl) {
        margin-right: 24px;
      }
    }

    @include mq(md) {
      flex: 2;
      max-width: none;
      padding: 80px;
    }

    @include mq(xl) {
      flex-direction: row;
      align-items: center;
      background-image: url("assets/image/tv_sample.png");
      animation-name: opacity;
      animation-duration: 1s;
      animation-timing-function: cubic-bezier(1, 1, 1, 1);
      background-repeat: no-repeat;
      background-size: 100% 100%;
      padding: 10% 4% 8% 6%;
    }

    @include mq(smHeight) {
      padding: 7% 4% 6% 6%;
    }

    @include mq(xxl) {
      padding-left: 10%;
    }
  }

  &-scroll-wrapper {
    height: 100%;
    overflow: scroll;
    display: flex;
    flex-direction: column;

    @include mq(xl) {
      flex-direction: row;
      align-items: center;
    }
  }

  &-arrow {
    width: fit-content;
    display: flex;
    white-space: nowrap;
    align-items: center;
    gap: 16px;
    cursor: pointer;
    margin-bottom: 16px;
    opacity: 0.6;

    > img {
      width: 32px;
    }

    &:hover {
      opacity: 1;
    }

    @include mq(md) {
      display: none;
    }
  }

  &-detail-container {
    display: flex;
    flex-direction: column;
    gap: 16px;

    > ul {
      gap: 16px;
    }

    @include mq(xl) {
      width: 40%;
    }
  }

  &-title-wrapper {
    display: flex;
    justify-content: space-between;

    > button {
      height: fit-content;
      border-radius: 24px;
      padding: 12px 16px;
      cursor: pointer;
      white-space: nowrap;
      background-color: cornflowerblue;
      color: white;
      border: 1px solid cornflowerblue;
      &:hover {
        opacity: 0.8;
      }
      &:active {
        background-color: white;
        color: lightslategrey;
      }

      &.remove {
        background: lightcoral;
      }
    }
  }
}
