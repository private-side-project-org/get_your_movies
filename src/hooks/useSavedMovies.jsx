import React, { createContext, useContext, useState } from "react";
import PropTypes from "prop-types";
const SavedMovieContext = createContext({});

const savedMovieContextValues = () => {
  const [favoriteMovieList, setFavoriteMovieList] = useState(
    JSON.parse(localStorage.getItem("movies"))
  );
  const getFavoriteMovie = (movie) =>
    favoriteMovieList.find((favMovie) => favMovie.id === movie.id);

  return {
    favoriteMovieList,
    setFavoriteMovieList,
    getFavoriteMovie,
  };
};

export const SavedMoviesProvider = ({ children }) => {
  return (
    <SavedMovieContext.Provider value={savedMovieContextValues()}>
      {children}
    </SavedMovieContext.Provider>
  );
};

SavedMoviesProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default () => useContext(SavedMovieContext);
