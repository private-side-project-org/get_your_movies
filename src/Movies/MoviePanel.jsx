import React from "react";
import PropTypes from "prop-types";
import useGetMovieById from "queries/useGetMovieById";
import getMovieImage from "utils/getMovieImage";
import ClipLoader from "react-spinners/ClipLoader";
import { toast } from "react-toastify";

import "./moviePanel.scss";

const arrow = require("assets/icons/arrow.svg");
const noImageAvailable = require("assets/icons/camera-reel.svg");

const propTypes = {
  selectedMovie: PropTypes.shape({
    id: PropTypes.number,
  }).isRequired,
  onSetSelectedMovie: PropTypes.func.isRequired,
  selectedTab: PropTypes.string.isRequired,
};

const loaderStyle = {
  position: "absolute",
  top: "45%",
  left: "40%",
};

const MoviePanel = ({ selectedMovie, onSetSelectedMovie, selectedTab }) => {
  const { movie, isLoading, isFetching } = useGetMovieById(selectedMovie.id, {
    enabled: selectedTab !== "favorite",
  });

  const displayMovie = selectedTab !== "favorite" ? movie : selectedMovie;
  console.log("displaymovie", displayMovie);

  const poster = displayMovie?.poster_path
    ? getMovieImage(displayMovie?.poster_path, 400)
    : noImageAvailable;

  const handleSaveMovie = () => {
    const savedMovies = JSON.parse(localStorage.getItem("movies")) || [];
    const hasSaved = savedMovies?.some(
      (savedMovie) => savedMovie.id === movie.id
    );

    if (!hasSaved) {
      localStorage.setItem(
        "movies",
        JSON.stringify([...savedMovies, { ...movie }])
      );
      onSetSelectedMovie(movie);
      toast.success("Movie has been saved");
    } else {
      toast.error("You've already saved movie");
    }
  };

  return (
    <div className="moviePanel-container">
      {isLoading || isFetching || !poster ? (
        <ClipLoader color="#000000" css={loaderStyle} size={80} />
      ) : (
        <>
          <div className="moviePanel-image-container">
            <div
              className="moviePanel-arrow"
              onClick={() => onSetSelectedMovie(null)}
            >
              <img src={arrow} alt="back_to_list_arrow" />
              <p>Back to the list</p>
            </div>
            <img src={poster} alt="poster" />
          </div>
          <div className="moviePanel-detail-container">
            <div className="moviePanel-title-wrapper">
              <h2>{displayMovie.title}</h2>
              {selectedTab !== "favorite" && (
                <button type="button" onClick={handleSaveMovie}>
                  Save movie
                </button>
              )}
            </div>
            <div className="flex-row">
              <h4>Genre</h4>
              <ul className="flex-row">
                {displayMovie?.genres.map((genre) => {
                  return (
                    <li key={genre.id} className="label">
                      {genre.name}
                    </li>
                  );
                })}
              </ul>
            </div>
            <div className="flex-row">
              <h4>Language</h4>
              <ul className="flex-row">
                {displayMovie?.spoken_languages.map((lang) => {
                  return (
                    <li key={lang.english_name} className="label">
                      {lang.name}
                    </li>
                  );
                })}
              </ul>
            </div>
            <p>{displayMovie.overview}</p>
            <div className="flex-row">
              <h4>Productions:</h4>
              <ul className="flex-column">
                {displayMovie.production_companies.map((company) => {
                  return <li key={company.id}>{company.name}</li>;
                })}
              </ul>
            </div>
          </div>
        </>
      )}
    </div>
  );
};

MoviePanel.propTypes = propTypes;

export default MoviePanel;
