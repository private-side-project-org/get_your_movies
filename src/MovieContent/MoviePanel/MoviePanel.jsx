import React from "react";
import PropTypes from "prop-types";
import useGetMovieById from "queries/useGetMovieById";
import getMovieImage from "utils/getMovieImage";
import ClipLoader from "react-spinners/ClipLoader";
import { toast } from "react-toastify";
import useSavedMovies from "hooks/useSavedMovies";
import CONSTANTS from "utils/constants";

import "./moviePanel.scss";

const { MOVIES } = CONSTANTS.LOCAL_STORAGE_KEYS;

const arrow = require("assets/icons/arrow.svg");
const noImageAvailable = require("assets/icons/camera-reel.svg");

const propTypes = {
  selectedMovie: PropTypes.shape({
    id: PropTypes.number,
  }).isRequired,
  onSetSelectedMovie: PropTypes.func.isRequired,
};

const loaderStyle = {
  position: "absolute",
  top: "45%",
  left: "40%",
};

const MoviePanel = ({ selectedMovie, onSetSelectedMovie }) => {
  const { favoriteMovieList, getFavoriteMovie, setFavoriteMovieList } =
    useSavedMovies();

  // single movie to be displayed on panel if exists
  const favoriteMovie = getFavoriteMovie(selectedMovie);

  // query to get single movie to be displayed if favorite doesn't exist
  const { movie, isLoading, isFetching } = useGetMovieById(selectedMovie.id, {
    enabled: !favoriteMovie,
  });

  // define which movie(favorite or movie from query) to be displayed
  const displayMovie = favoriteMovie || movie;

  const handleSaveMovie = () => {
    const hasSaved = favoriteMovieList?.some(
      (savedMovie) => savedMovie.id === movie.id
    );

    if (!hasSaved) {
      localStorage.setItem(
        MOVIES,
        JSON.stringify([...favoriteMovieList, { ...movie }])
      );
      setFavoriteMovieList([...favoriteMovieList, { ...movie }]);
      onSetSelectedMovie(movie);
      toast.success("Movie has been saved");
    } else {
      toast.error("You've already saved movie");
    }
  };

  const handleRemoveMovie = () => {
    // get index of the movie to be removed
    const removeMovieIndex = favoriteMovieList.findIndex(
      (movie) => movie.id === selectedMovie.id
    );
    // take out target movie from favorite list
    favoriteMovieList.splice(removeMovieIndex, 1);
    // set updated movie on context
    setFavoriteMovieList(favoriteMovieList);
    // set updated movie list on local storage
    localStorage.setItem("movies", JSON.stringify(favoriteMovieList));

    // set id on selectedMovie to run query after remove movie in order to persist to display movie on panel
    onSetSelectedMovie({ id: selectedMovie.id });
    toast.success("Movie has been removed");
  };

  const poster = displayMovie?.poster_path
    ? getMovieImage(displayMovie?.poster_path, 400)
    : noImageAvailable;

  return (
    <div className="moviePanel-container">
      <div className="moviePanel-scroll-wrapper">
        {isLoading || isFetching ? (
          <ClipLoader color="#000000" css={loaderStyle} size={80} />
        ) : (
          <>
            <div className="moviePanel-image-container">
              <div
                className="moviePanel-arrow"
                onClick={() => onSetSelectedMovie(null)}
              >
                <img src={arrow} alt="back_to_list_arrow" />
                <p>Back to the list</p>
              </div>
              <img src={poster} alt="poster" />
            </div>
            <div className="moviePanel-detail-container">
              <div className="moviePanel-title-wrapper">
                <h2>{displayMovie.title}</h2>
                {
                  <button
                    type="button"
                    onClick={
                      favoriteMovie ? handleRemoveMovie : handleSaveMovie
                    }
                    className={`${favoriteMovie ? "remove" : ""}`}
                  >
                    {favoriteMovie ? "Remove" : "Save movie"}
                  </button>
                }
              </div>
              <div className="flex-row">
                <h4>Genre</h4>
                <ul className="flex-row">
                  {displayMovie?.genres.map((genre) => {
                    return (
                      <li key={genre.id} className="label">
                        {genre.name}
                      </li>
                    );
                  })}
                </ul>
              </div>
              <div className="flex-row">
                <h4>Language</h4>
                <ul className="flex-row">
                  {displayMovie?.spoken_languages.map((lang) => {
                    return (
                      <li key={lang.english_name} className="label">
                        {lang.name}
                      </li>
                    );
                  })}
                </ul>
              </div>
              <p>{displayMovie.overview}</p>
              <div className="flex-row">
                <h4>Productions:</h4>
                <ul className="flex-column">
                  {displayMovie.production_companies.map((company) => {
                    return <li key={company.id}>{company.name}</li>;
                  })}
                </ul>
              </div>
            </div>
          </>
        )}
      </div>
    </div>
  );
};

MoviePanel.propTypes = propTypes;

export default MoviePanel;
