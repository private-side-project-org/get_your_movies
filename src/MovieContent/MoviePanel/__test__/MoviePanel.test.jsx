import React from "react";
import PropTypes from "prop-types";
import { screen, fireEvent, render } from "@testing-library/react";
import MoviePanel from "../MoviePanel";
import { SavedMovieProvider } from "hooks/useSavedMovies";
import { QueryClientProvider, QueryClient } from "react-query";
import { ToastContainer } from "react-toastify";

const mockedMovie = {
  title: "test title",
  genres: [
    { id: 1, name: "genre1" },
    { id: 2, name: "genre2" },
  ],
  spoken_languages: [{ english_name: "test english name", name: "en" }],
  overview: "test overview",
  production_companies: [
    { id: 1, name: "test company1" },
    { id: 2, name: "test company2" },
  ],
};

jest.mock("../../../queries/useGetMovieById", () =>
  jest.fn(() => ({ movie: mockedMovie, isLoading: false, isFetching: false }))
);

const client = new QueryClient();

const WrapperComponent = ({ children }) => (
  <QueryClientProvider client={client}>
    <SavedMovieProvider>{children}</SavedMovieProvider>
    <ToastContainer />
  </QueryClientProvider>
);

WrapperComponent.propTypes = {
  children: PropTypes.node.isRequired,
};

describe("MoviePanel component test", () => {
  it("renders and snapshot component", () => {
    const props = {
      selectedMovie: {},
      onSetSelectedMovie: jest.fn(),
    };

    const renederedComponent = render(<MoviePanel {...props} />, {
      wrapper: WrapperComponent,
    });

    expect(renederedComponent.baseElement).toMatchSnapshot();
  });

  it("clicks action button and save movie", async () => {
    const props = {
      selectedMovie: {},
      onSetSelectedMovie: jest.fn(),
    };

    render(<MoviePanel {...props} />, {
      wrapper: WrapperComponent,
    });

    const actionButton = await screen.findByTestId("action_button");
    fireEvent.click(actionButton);

    expect(await screen.findByText("Movie has been saved")).toBeDefined();
  });

  it("click to back to list", async () => {
    const props = {
      selectedMovie: {},
      onSetSelectedMovie: jest.fn(),
    };

    render(<MoviePanel {...props} />, {
      wrapper: WrapperComponent,
    });

    const backToListArrow = await screen.findByTestId("back_to_list_arrow");
    fireEvent.click(backToListArrow);

    expect(props.onSetSelectedMovie).toHaveBeenCalled();
  });

  it("clicks action button and remove movie", async () => {
    const mockedGetFavoriteMovie = jest.fn(() => mockedMovie);
    const mockedFavoriteMovieList = [];
    const mockedSetFavoriteMovieList = jest.fn(() => {});

    const mockedHooks = jest.fn(() => ({
      favoriteMovieList: mockedFavoriteMovieList,
      getFavoriteMovie: mockedGetFavoriteMovie,
      setFavoriteMovieList: mockedSetFavoriteMovieList,
    }));

    React.useContext = mockedHooks;

    const props = {
      selectedMovie: {},
      onSetSelectedMovie: jest.fn(),
    };

    render(<MoviePanel {...props} />, {
      wrapper: WrapperComponent,
    });

    const actionButton = await screen.findByTestId("action_button");
    fireEvent.click(actionButton);

    expect(await screen.findByText("Movie has been removed")).toBeDefined();
  });
});
