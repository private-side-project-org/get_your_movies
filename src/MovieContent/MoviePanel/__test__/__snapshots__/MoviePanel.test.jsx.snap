// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviePanel component test renders and generate snapshot 1`] = `
<body>
  <div>
    <div
      class="moviePanel-container"
    >
      <div
        class="moviePanel-scroll-wrapper"
      >
        <div
          class="moviePanel-image-container"
        >
          <div
            class="moviePanel-arrow"
            data-testid="back_to_list_arrow"
          >
            <img
              alt="back_to_list_arrow"
              src="[object Object]"
            />
            <p>
              Back to the list
            </p>
          </div>
          <img
            alt="poster"
            src="[object Object]"
          />
        </div>
        <div
          class="moviePanel-detail-container"
        >
          <div
            class="moviePanel-title-wrapper"
          >
            <h2>
              test title
            </h2>
            <button
              class=""
              data-testid="action_button"
              type="button"
            >
              Save movie
            </button>
          </div>
          <div
            class="flex-row"
          >
            <h4>
              Genre
            </h4>
            <ul
              class="flex-row moviePanel-genre-wrapper"
            >
              <li
                class="label"
              >
                genre1
              </li>
              <li
                class="label"
              >
                genre2
              </li>
            </ul>
          </div>
          <div
            class="flex-row"
          >
            <h4>
              Language
            </h4>
            <ul
              class="flex-row"
            >
              <li
                class="label"
              >
                en
              </li>
            </ul>
          </div>
          <p>
            test overview
          </p>
          <div
            class="flex-row"
          >
            <h4>
              Productions:
            </h4>
            <ul
              class="flex-column"
            >
              <li>
                test company1
              </li>
              <li>
                test company2
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
    <div
      class="Toastify"
    />
  </div>
</body>
`;
